README.md
IRC protocol support for the XP Framework
========================================================================
 
Usage example:

```php
use peer\irc\IRCConnection;
use util\log\Logger;
use util\log\ConsoleAppender;

$c= new IRCConnection(new IRCUser('MyBot'), 'irc.xxx.net');

// Add logging
$cat= Logger::getInstance()->getCategory();
$cat->addAppender(new ConsoleAppender());
$c->setTrace($cat);

// Add listener
$c->addListener(new MyBotListener());
$c->open();
$c->run();
$c->close();
```

The `MyBotListener` class should subclass the abstract `IRCConnectionListener`
base class, overwriting one or more of the following template methods:

```groovy
public abstract class peer.irc.IRCConnectionListener extends lang.Object {
  public void onPings(peer.irc.IRCConnection $connection, string $data)
  public void onConnect(peer.irc.IRCConnection $connection, string $server, int $port)
  public void onDisconnect(peer.irc.IRCConnection $connection, string $server, int $port)
  public void onMOTDStart(peer.irc.IRCConnection $connection, string $server, string $target, string $data)
  public void onMOTD(peer.irc.IRCConnection $connection, string $server, string $target, string $data)
  public void onEndOfMOTD(peer.irc.IRCConnection $connection, string $server, string $target, string $data)
  public void onServerMessage(peer.irc.IRCConnection $connection, string $server, int $code, string $target, string $data)
  public void onInvite(peer.irc.IRCConnection $connection, string $nick, string $who, string $channel)
  public void onKicks(peer.irc.IRCConnection $connection, string $channel, string $nick, string $who, string $reason)
  public void onQuits(peer.irc.IRCConnection $connection, string $channel, string $nick, string $reason)
  public void onNickChanges(peer.irc.IRCConnection $connection, string $nick, string $new)
  public void onJoins(peer.irc.IRCConnection $connection, string $channel, string $nick)
  public void onParts(peer.irc.IRCConnection $connection, string $channel, string $nick, string $message)
  public void onModeChanges(peer.irc.IRCConnection $connection, string $nick, string $target, string $mode, string $params)
  public void onPrivateMessage(peer.irc.IRCConnection $connection, string $nick, string $target, string $message)
  public void onTopic(peer.irc.IRCConnection $connection, string $nick, string $channel, string $topic)
  public void onNotice(peer.irc.IRCConnection $connection, string $nick, string $target, string $message)
  public void onAction(peer.irc.IRCConnection $connection, string $nick, string $target, string $action)
  public void onVersion(peer.irc.IRCConnection $connection, string $nick, string $target, string $params)
  public void onUserInfo(peer.irc.IRCConnection $connection, string $nick, string $target, string $params)
  public void onClientInfo(peer.irc.IRCConnection $connection, string $nick, string $target, string $params)
  public void onPing(peer.irc.IRCConnection $connection, string $nick, string $target, string $params)
  public void onTime(peer.irc.IRCConnection $connection, string $nick, string $target, string $params)
  public void onFinger(peer.irc.IRCConnection $connection, string $nick, string $target, string $params)
  public string hashCode()
  public bool equals(lang.Generic $cmp)
  public string getClassName()
  public lang.XPClass getClass()
  public string toString()
}
```

ChangeLog.md
IRC protocol support for the XP Framework ChangeLog
========================================================================

## ?.?.? / ????-??-??

* Heads up: Converted classes to PHP 5.3 namespaces - (@thekid)



src/main/php/peer/irc/IRCUser.class.php
<?php namespace peer\irc;

/**
 * Represents an IRC User
 *
 * @see   xp://peer.irc.IRCConnection
 */
class IRCUser extends \lang\Object {
  public
    $nick     = '',
    $username = '',
    $hostname = '',
    $realname = '';
    
  /**
   * Constructor
   *
   * @param   string $nick
   * @param   string $realname default NULL (defaults to nickname)
   * @param   string $username default NULL (defaults to current username)
   * @param   string $hostname default 'localhost'
   */
  public function __construct($nick, $realname= null, $username= null, $hostname= 'localhost') {
    $this->nick= $nick;
    $this->realname= $realname ? $realname : $nick;
    $this->username= $username ? $username : get_current_user();
    $this->hostname= $hostname;
  }

  /**
   * Set Nick
   *
   * @param   string $nick
   */
  public function setNick($nick) {
    $this->nick= $nick;
  }

  /**
   * Get Nick
   *
   * @return  string
   */
  public function getNick() {
    return $this->nick;
  }

  /**
   * Set Username
   *
   * @param   string $username
   */
  public function setUsername($username) {
    $this->username= $username;
  }

  /**
   * Get Username
   *
   * @return  string
   */
  public function getUsername() {
    return $this->username;
  }

  /**
   * Set Hostname
   *
   * @param   string $hostname
   */
  public function setHostname($hostname) {
    $this->hostname= $hostname;
  }

  /**
   * Get Hostname
   *
   * @return  string
   */
  public function getHostname() {
    return $this->hostname;
  }

  /**
   * Set Realname
   *
   * @param   string $realname
   */
  public function setRealname($realname) {
    $this->realname= $realname;
  }

  /**
   * Get Realname
   *
   * @return  string
   */
  public function getRealname() {
    return $this->realname;
  }
}


src/main/php/peer/irc/IRCColor.class.php
<?php namespace peer\irc;

define('IRC_COLOR_DEFAULT',     0);
define('IRC_COLOR_BLACK',       1);
define('IRC_COLOR_DARKBLUE',    2);
define('IRC_COLOR_DARKGREEN',   3);
define('IRC_COLOR_RED',         4);
define('IRC_COLOR_PURPLE',      5);
define('IRC_COLOR_DARKRED',     6);
define('IRC_COLOR_ORANGE',      7);
define('IRC_COLOR_YELLOW',      8);
define('IRC_COLOR_GREEN',       9);
define('IRC_COLOR_MAGENTA',    10);
define('IRC_COLOR_STEELBLUE',  11);
define('IRC_COLOR_BLUE',       12);
define('IRC_COLOR_PINK',       13);
define('IRC_COLOR_DARKGRAY',   14);
define('IRC_COLOR_LIGHTGRAY',  15);
define('IRC_COLOR_WHITE',      16);


/**
 * IRC colors
 *
 * Example:
 * <code>
 *   $connection->sendMessage(
 *     '#test', 
 *     '%sThis is red', 
 *     IRCColor::forCode(IRC_COLOR_RED)
 *   );
 * </code>
 *
 * @purpose  Utility class
 */
class IRCColor extends \lang\Object {

  /**
   * Retrieves color representation
   *
   * @param   int code
   * @return  string color representation
   */
  public static function forCode($code) {
    return "\3".$code;
  }
}


src/main/php/peer/irc/IRCConnectionListener.class.php
<?php namespace peer\irc;

/**
 * Connection listener
 *
 * @see    xp://peer.irc.IRCConnection#addListener
 */
abstract class IRCConnectionListener extends \lang\Object {

  /**
   * Callback for Pings. Note that the PING has already been answered
   * when this method is called, so you won't have to send a PONG 
   * yourself.
   *
   * You might want to use this method in IRC-bots to accomplish the
   * task of deliberately being able to perform an action without any
   * other action having taken place (e.g., maintenance, reload config,
   * ...)
   *
   * @param   peer.irc.IRCConnection connection
   * @param   string data
   * @return  void
   */
  public function onPings($connection, $data) { }

  /**
   * Callback for when a connection to the IRC server has been 
   * established. This method is called *after* a connecting was
   * successful.
   *
   * @param   peer.irc.IRCConnection connection
   * @param   string server
   * @param   int port
   * @return  void
   */
  public function onConnect($connection, $server, $port) { }

  /**
   * Callback for when a connection to the IRC server has been 
   * closed. This method is called *before* the connection is actually
   * dropped; thus making it possible to say goodbye. You cannot do
   * anything to prevent disconnection, though.
   *
   * @param   peer.irc.IRCConnection connection
   * @param   string server
   * @param   int port
   * @return  void
   */
  public function onDisconnect($connection, $server, $port) { }
  
  /**
   * Callback for server message MOTDSTART (375)
   *
   * @param   peer.irc.IRCConnection connection
   * @param   string server
   * @param   string target whom the message is for
   * @param   string data
   * @return  void
   */
  public function onMOTDStart($connection, $server, $target, $data) { }

  /**
   * Callback for server message MOTD (372)
   *
   * @param   peer.irc.IRCConnection connection
   * @param   string server
   * @param   string target whom the message is for
   * @param   string data
   * @return  void
   */
  public function onMOTD($connection, $server, $target, $data) { }

  /**
   * Callback for server message REPLY_ENDOFMOTD (376)
   *
   * @param   peer.irc.IRCConnection connection
   * @param   string server
   * @param   string target whom the message is for
   * @param   string data
   * @return  void
   */
  public function onEndOfMOTD($connection, $server, $target, $data) { }
  
  /**
   * Callback for all other server messages
   *
   * @param   peer.irc.IRCConnection connection
   * @param   string server
   * @param   int code one of the IRC_* constants from peer.irc.IRCConstants
   * @param   string target whom the message is for
   * @param   string data
   * @return  void
   */
  public function onServerMessage($connection, $server, $code, $target, $data) { }

  /**
   * Callback for invitations. Note: Due to the limitations of the INVITE
   * command, you won't be able to join password-protected channels unless
   * you know their password!
   *
   * Example: Join if we're invited:
   * 
   * ```php
   * $connection->join($channel);
   * ```
   *
   * @param   peer.irc.IRCConnection connection
   * @param   string nick sending the invitation
   * @param   string who who is invited
   * @param   string channel invitation is for
   * @return  void
   */
  public function onInvite($connection, $nick, $who, $channel) { }

  /**
   * Callback for kicks
   *
   * Example (identifying being kicked):
   * 
   * ```php
   * if (strcasecmp($who, $connection->user->getNick()) == 0) {
   *   // ... I was kicked ...
   * }
   * ```
   *
   * @param   peer.irc.IRCConnection connection
   * @param   string channel the channel the user was kicked from
   * @param   string nick that initiated the kick
   * @param   string who who was kicked
   * @param   string reason what reason the user was kicked for
   * @return  void
   */
  public function onKicks($connection, $channel, $nick, $who, $reason) { }

  /**
   * Callback for quits
   *
   * @param   peer.irc.IRCConnection connection
   * @param   string channel the channel the user quit from
   * @param   string nick who quit
   * @param   string reason what reason the user supplied for quitting
   * @return  void
   */
  public function onQuits($connection, $channel, $nick, $reason) { }

  /**
   * Callback for nick changes
   *
   * @param   peer.irc.IRCConnection connection
   * @param   string nick the old nick
   * @param   string new the new nick
   * @return  void
   */
  public function onNickChanges($connection, $nick, $new) { }

  /**
   * Callback for joins
   *
   * Example (welcome users)
   * 
   * ```php
   * // Send it to the channel so everybody knows
   * $connection->sendMessage($channel, 'Welcome %s', $nick);
   *
   * // Send it to the joinee privately
   * $connection->sendMessage($nick, 'Welcome!');
   * ```
   *
   * @param   peer.irc.IRCConnection connection
   * @param   string channel which channel was joined
   * @param   string nick who joined
   * @return  void
   */
  public function onJoins($connection, $channel, $nick) { }

  /**
   * Callback for parts
   *
   * @param   peer.irc.IRCConnection connection
   * @param   string channel which channel was part
   * @param   string nick who part
   * @param   string message the part message, if any
   * @return  void
   */
  public function onParts($connection, $channel, $nick, $message) { }
  
  /**
   * Callback for mode changes
   *
   * @param   peer.irc.IRCConnection connection
   * @param   string nick who initiated the mode change
   * @param   string target what the mode setting is for (e.g. +k #channel, +i user)
   * @param   string mode the mode including a + or - as its first letter
   * @param   string params additional parameters
   * @return  void
   */
  public function onModeChanges($connection, $nick, $target, $mode, $params) { }

  /**
   * Callback for private messages
   *
   * Example (implementing "commands"):
   *
   * ```php
   * if (sscanf($message, "!%s %[^\r]", $command, $params)) {
   *   switch (strtolower($command)) {
   *     case 'status':
   *       // ...
   *   }
   * }
   * ```
   *
   * @param   peer.irc.IRCConnection connection
   * @param   string nick
   * @param   string target
   * @param   string message
   * @return  void
   */
  public function onPrivateMessage($connection, $nick, $target, $message) { }

  /**
   * Callback for topic changes
   *
   * @param   peer.irc.IRCConnection connection
   * @param   string nick who changed the topic
   * @param   string channel what channel the topic was changed for
   * @param   string topic the new topic
   * @return  void
   */
  public function onTopic($connection, $nick, $channel, $topic) { }

  /**
   * Callback for notices
   *
   * @param   peer.irc.IRCConnection connection
   * @param   string nick
   * @param   string target
   * @param   string message
   * @return  void
   */
  public function onNotice($connection, $nick, $target, $message) { }

  /**
   * Callback for actions. Actions are when somebody writes /me ...
   * in their IRC window.
   *
   * Example (annoying:)):
   * 
   * ```php
   * $connection->sendAction($target, 'imitates %s and %s, too', $nick, $params);
   * ```
   *
   * @param   peer.irc.IRCConnection connection
   * @param   string nick who initiated the action
   * @param   string target where action was initiated
   * @param   string action what actually happened (e.g. "looks around")
   * @return  void
   */
  public function onAction($connection, $nick, $target, $action) { }

  /**
   * Callback for CTCP VERSION
   *
   * @param   peer.irc.IRCConnection connection
   * @param   string nick nick requesting version
   * @param   string target where version was requested
   * @param   string params additional parameters
   * @return  void
   */
  public function onVersion($connection, $nick, $target, $params) { }

  /**
   * Callback for CTCP USERINFO
   *
   * @param   peer.irc.IRCConnection connection
   * @param   string nick nick requesting user information
   * @param   string target where user information was requested
   * @param   string params additional parameters
   * @return  void
   */
  public function onUserInfo($connection, $nick, $target, $params) { }

  /**
   * Callback for CTCP CLIENTINFO
   *
   * @param   peer.irc.IRCConnection connection
   * @param   string nick nick requesting client information
   * @param   string target where client information was requested
   * @param   string params additional parameters
   * @return  void
   */
  public function onClientInfo($connection, $nick, $target, $params) { }

  /**
   * Callback for CTCP PING
   *
   * @param   peer.irc.IRCConnection connection
   * @param   string nick nick requesting ping
   * @param   string target where ping was requested
   * @param   string params additional parameters
   * @return  void
   */
  public function onPing($connection, $nick, $target, $params) { }

  /**
   * Callback for CTCP TIME
   *
   * @param   peer.irc.IRCConnection connection
   * @param   string nick nick requesting time
   * @param   string target where time was requested
   * @param   string params additional parameters
   * @return  void
   */
  public function onTime($connection, $nick, $target, $params) { }

  /**
   * Callback for CTCP FINGER
   *
   * @param   peer.irc.IRCConnection connection
   * @param   string nick nick requesting finger information
   * @param   string target where finger information was requested
   * @param   string params additional parameters
   * @return  void
   */
  public function onFinger($connection, $nick, $target, $params) { }

}


src/main/php/peer/irc/IRCConnection.class.php
<?php namespace peer\irc;

use peer\Socket;
use util\log\Traceable;

/**
 * IRC Connection
 *
 * Usage example:
 * <code>
 *   uses(
 *     'peer.irc.IRCConnection', 
 *     'util.log.Logger',
 *     'util.log.ConsoleAppender',
 *     'KrokerdilBotListener'
 *   );
 *   
 *   $c= new IRCConnection(new IRCUser('KrokerdilBot'), 'irc.xxx.net');
 *   
 *   $cat= Logger::getInstance()->getCategory();
 *   $cat->addAppender(new ConsoleAppender());
 *   $c->setTrace($cat);
 *   
 *   $c->addListener(new KrokerdilBotListener());
 *   try {
 *     $c->open();
 *     $c->run();
 *     $c->close();
 *   } catch (XPException $e) {
 *     $e->printStackTrace();
 *   }
 * </code>
 *
 * @see      xp://peer.irc.IRCConstants
 * @see      xp://peer.irc.IRCUser
 * @purpose  A socket connection to a RFC-1459 compatible IRC server.
 */
class IRCConnection extends \lang\Object implements Traceable {
  public
    $sock         = null,
    $cat          = null,
    $user         = null,
    $listeners    = array();

  static function __static() {
    \lang\XPClass::forName('peer.irc.IRCConstants');  // TODO: Use class constants
  }

  /**
   * Constructor
   *
   * @param   peer.irc.IRCUser $user
   * @param   string $server
   * @param   int $port default 6667
   */
  public function __construct($user, $server, $port= 6667) {
    $this->user= $user;
    $this->sock= new Socket($server, $port);
  }
  
  /**
   * Adds a listener
   *
   * @see     xp://peer.irc.IRCConnectionListener
   * @param   peer.irc.IRCConnectionListener $listener
   * @return  peer.irc.IRCConnectionListener the listener added
   */
  public function addListener($listener) {
    $this->listeners[]= $listener;
    return $listener;
  }
  
  /**
   * Set a logger category for debugging
   *
   * @param   util.log.LogCategory $cat
   */
  public function setTrace($cat) {
    $this->cat= $cat;
  }

  /**
   * Opens connection to the IRC server. Also sets the socket to 
   * blocking mode.
   *
   * @see     xp://peer.Socket#connect
   * @see     xp://peer.Socket#setBlocking
   * @throws  peer.SocketException in case connecting fails
   */
  public function open() {
    if (!$this->sock->connect()) return false;
    $this->sock->setBlocking(true);
    $this->notify('connect', $this->sock->host, $this->sock->port); 
  }
  
  /**
   * Notify all listeners
   *
   * @param   string event
   * @param   var* parameters
   */
  protected function notify() {
    $a= func_get_args();
    $function= 'on'.ucfirst(array_shift($a));
    array_unshift($a, $this);
    for ($i= 0, $s= sizeof($this->listeners); $i < $s; $i++) {
      $this->cat && $this->cat->debugf(
        'Calling %s::%s(<self>, %s)', 
        $this->listeners[$i]->getClassName(), 
        $function, 
        \xp::stringOf(array_slice($a, 1))
      );
      call_user_func_array(array($this->listeners[$i], $function), $a);
    }
  }
  
  /**
   * Process a single line of read data
   *
   * @param   string r the data
   * @return  bool TRUE when line was successfully processed
   */
  public function process($r) { 
    if (':' == $r{0}) {
    
      if (4 == sscanf($r, ":%s %d %s %[^\r]", $server, $code, $target, $data)) {

        // Server messages
        // :irc.xxx.net 372 KrokerdilBot :- Excuse of the hour:
        // :irc.xxx.net 353 KrokerdilBot = #test :KrokerdilBot @krokerdil
        // :irc.xxx.net 366 KrokerdilBot #test :End of /NAMES list.
        // :irc.xxx.net 376 KrokerdilBot :End of /MOTD command.
        $this->cat && $this->cat->debugf('    Server message #%d %s', $code, IRCConstants::nameOf($code));
        
        switch ($code) {
          case IRC_REPLY_MOTDSTART:
          case IRC_REPLY_MOTD:
          case IRC_REPLY_ENDOFMOTD:
            $this->notify(substr(IRCConstants::nameOf($code), 6), $server, $target, $data);
            break;
          
          default:
            $this->notify('serverMessage', $server, $code, $target, $data);
        }
      } else if (4 <= sscanf($r, ":%[^!]!%s %s %s %[^\r]", $nick, $user, $action, $target, $data)) {

        // Other messages
        switch (strtolower($action)) {

          // :krokerdil!thekid@xxx.de PRIVMSG KrokerdilBot :Wop
          // :krokerdil!thekid@xxx.de PRIVMSG #test :Poff
          // :krokerdil!thekid@xxx.de PRIVMSG #test :\001ACTION thinks Bot is stoopid\001
          // :krokerdil!thekid@xxx.de PRIVMSG Bot :\001USERINFO\001
          // :krokerdil!thekid@xxx.de PRIVMSG KrokerdilBot :Hello\0033Hello
          case 'privmsg':
            if ("\1" == $data{1}) {
              list($command, $params)= explode(' ', substr($data, 2, -1), 2);
              $this->notify($command, $nick, $target, $params);
            } else {
              $this->notify('privateMessage', $nick, $target, substr($data, 1));
            }
            break;

          // :krokerdil!thekid@xxx.de JOIN :#test
          case 'join':
            $this->notify('joins', substr($target, 1), $nick);
            break;

          // :krokerdil!thekid@xxx.de PART #test :Leaving
          case 'part':
            $this->notify('parts', $target, $nick, substr($data, 1));
            break;

          // :krokerdil!thekid@xxx.de MODE #test +v KrokerdilBot
          // :krokerdil!thekid@xxx.de MODE #test +k HALLO
          // :krokerdil!thekid@xxx.de MODE #test -k HALLO
          // :krokerdil!thekid@xxx.de MODE #test +b *!*thekid@*.xxx.de
          // :krokerdil!thekid@xxx.de MODE #test +o KrokerdilBot 
          case 'mode':
            list($mode, $params)= explode(' ', $data, 2);
            $this->notify('modeChanges', $nick, $target, $mode, $params);
            break;

          // :krokerdil!thekid@xxx.de KICK #schlund Bot :IHATEYOU
          case 'kick':
            list($who, $reason)= explode(' ', $data, 2);
            $this->notify('kicks', $target, $nick, $who, substr($reason, 1));
            break;

          // :krokerdil!thekid@xxx.de QUIT :signoff: Ave
          case 'quit':
            $this->notify('quits', $target, $nick, $data);
            break;            

          // :krokerdil!thekid@xxx.de NICK :schnutepanzerdil
          case 'nick':
            $this->notify('nickChanges', $nick, substr($target, 1));
            break;
          
          // :krokerdil!thekid@xxx.de INVITE KrokerdilBot :#test
          case 'invite':
            $this->notify('invite', $nick, $target, substr($data, 1));
            break;
          
          // :krokerdil!thekid@xxx.de NOTICE #test :8tung baby
          // :krokerdil!thekid@xxx.de NOTICE KrokerdilBot :8tung baby
          case 'notice':
            $this->notify('notice', $nick, $target, substr($data, 1));
            break;
            
          // :krokerdil!thekid@xxx.de TOPIC #test :Korkerdl             
          case 'topic':
            $this->notify('topic', $nick, $target, substr($data, 1));
            break;

          default:
            $this->cat && $this->cat->warn('Unrecognized action', $action, '(', $nick, $user, $action, $target, $data, ')');
            return false;
        }
      } else {
        $this->cat && $this->cat->error('Could not parse', $r);
        return false;
      }
    } else {
      $this->cat && $this->cat->error('Malformed message', $r);
      return false;
    }
    return true;
  }
  
  /**
   * Set nickname and update this user's nickname.
   *
   * @param   string nick
   * @return  bool success TRUE in case the command could be sent
   * @throws  lang.IllegalStateException in case the socket is not connected
   */
  public function setNick($nick) {
    if (!$this->sock->isConnected()) {
      throw new \lang\IllegalStateException('Not connected');
    }
    if (!$this->writeln('NICK %s', $nick)) return false;
    $this->user->setNick($nick);
    return true;
  }
  
  /**
   * Join a channel
   *
   * @param   string channel without leading #
   * @param   string keyword default NULL
   * @return  bool success TRUE in case the command could be sent
   * @throws  lang.IllegalStateException in case the socket is not connected
   */
  public function join($channel, $keyword= null) {
    if (!$this->sock->isConnected()) {
      throw new \lang\IllegalStateException('Not connected');
    }
    
    // Be tolerant about channel parameter and rip off leading # if necessary
    return $this->writeln('JOIN #%s %s', ltrim($channel, '#'), $keyword);
  }
  
  /**
   * Leave a channel
   *
   * @param   string channel without leading #
   * @return  bool success TRUE in case the command could be sent
   * @throws  lang.IllegalStateException in case the socket is not connected
   */
  public function part($channel) {
    if (!$this->sock->isConnected()) {
      throw new \lang\IllegalStateException('Not connected');
    }
    
    // Be tolerant about channel parameter and rip off leading # if necessary
    return $this->writeln('PART #%s', ltrim($channel, '#'));
  }

  /**
   * Wrapper around writeln() that sends a PRIVMSG
   *
   * @param   string dest either a nick or a channel (beginning with "#")
   * @param   string format
   * @param   var* format_arguments
   * @return  bool success TRUE in case the command could be sent
   */
  public function sendMessage() {
    $a= func_get_args();
    $dest= array_shift($a);
    return $this->writeln('PRIVMSG %s :%s', $dest, vsprintf(array_shift($a), $a));
  }

  /**
   * Wrapper around writeln() that sends a PRIVMSG with an ACTION
   *
   * @param   string dest either a nick or a channel (beginning with "#")
   * @param   string format
   * @param   var* format_arguments
   * @return  bool success TRUE in case the command could be sent
   */
  public function sendAction() {
    $a= func_get_args();
    $dest= array_shift($a);
    return $this->writeln("PRIVMSG %s :\1ACTION %s\1", $dest, vsprintf(array_shift($a), $a));
  }
  
  /**
   * The main loop in which the socket is polled for data, notifying
   * the attached listeners of events.
   *
   * @throws  lang.IllegalStateException in case the socket is not connected
   * @throws  io.IOException in case registration failed
   */
  public function run() {
    if (!$this->sock->isConnected()) {
      throw new \lang\IllegalStateException('Not connected');
    }
    
    // Register with the server
    $this->writeln(
      'USER %s %s %s :%s', 
      $this->user->getUsername(),
      $this->user->getHostname(),
      $this->sock->host,
      $this->user->getRealname()
    );

    // Set nickname
    $this->setNick($this->user->getNick());
    
    // Loop while socket is not disconnected
    $messages= array();
    while (!$this->sock->eof()) {
      if (!$r= $this->readln()) continue;
      
      // Reply to PINGs immediately
      if ('PING :' == substr($r, 0, 6)) {
        $this->writeln('PONG %s', substr($r, 6));
        $this->notify('pings', substr($r, 6));
        continue;
      }

      // Process data
      $this->process($r);
    }
  }
  
  /**
   * Close the communication socket. Issues a QUIT before 
   *
   * @see     xp://peer.Socket#close
   * @return  bool success
   */
  public function close() {
    if ($this->sock->isConnected()) return true;
    
    $this->notify('disconnect', $this->sock->host, $this->sock->port); 
    
    // Politely say goodbye
    $this->writeln('QUIT');
    return $this->sock->close();
  }

  /**
   * Directly read a line off the socket. The returned line has any 
   * trailing CRLF trimmed off.
   *
   * @see     xp://peer.Socket#readLine
   * @return  string or FALSE to indicate failure
   */    
  public function readln() {
    $r= $this->sock->readLine(0x2000);
    $this->cat && $this->cat->infof(
      '<<< %s %3d: %s', 
      gettype($r), 
      strlen($r), 
      addcslashes($r, "\0..\37")
    );
    
    return $r;
  }
  
  /**
   * Directly write a string to the socket. A CRLF is added automatically 
   * at the string's end.
   *
   * Note: It seems as if 448 bytes is the maximum length of a single
   * string that can be written e.g. as a PRIVMSG (including the PRIVMSG
   * command and destination). This limit is not checked upon.
   *
   * @see     xp://peer.Socket#write
   * @param   string format
   * @param   var* formatargs
   * @return  bool success
   */
  public function writeln() {
    $a= func_get_args();
    $str= array_shift($a);
    $cmd= vsprintf($str, $a);
    
    $this->cat && $this->cat->infof(
      '>>> %s %3d: %s', 
      gettype($cmd), 
      strlen($cmd), 
      addcslashes($cmd, "\0..\37")
    );
    
    return $this->sock->write($cmd."\r\n");
  }
} 
